Modelling motions of binary star system

henrysky opened the pull request:
I have added support to model motions of binary star system as seen on the sky and their 6D astrometry (RA, DEC, PMRA, PMDEC, PLX, RV). 

So far I have only manually added Alpha Centauri A/B, Sirius A/B and 61 Cygni A/B. In order to model a binary system we need:

- $e$ - Eccentricity
- $a$ - Semi-major axis
- $\Omega$ - Longitude of the ascending node
- $\omega$ - Argument of periapsis
- $i$ - Inclination
- $P$ - Period
- $T$ - Periastron epoch
- Barycenter RA/DEC, PMRA/PMDEC, PLX, RV
- Barycenter data epoch
- HIP or Gaia ID of primary and secondary stars
- Primary and secondary star masses

At least for Alpha Centauri A/B and Sirius A/B, the motion on the sky as well as radial velocity matched really well to past observations (https://iopscience.iop.org/article/10.3847/1538-3881/abfaff/pdf, https://arxiv.org/pdf/1703.10625).

If there are other binary systems you want to model, I can try to find all the required orbital parameters and add them manually before we figure it out a way to model a large amount of binary stars if possible.

Fixes #353 

### Screenshots (if appropriate):

An example of Sirius A/B with parallax disabled to avoid confusion:

https://github.com/user-attachments/assets/1bb75ddf-1288-4de4-9d0d-8ccd2ac911cf

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update
- [ ] Housekeeping

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: MacOS 15.2
* Graphics Card: Apple M3

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

gzotti commented:
Great stuff! But probably #4076 should be 100% fixed before adding this.  

&gamma; Vir is another nice double. For a catalog, this looks promising: http://www.astro.gsu.edu/wds/orb6.html

alex-w commented:
> Great stuff! But probably #4076 should be 100% fixed before adding this.
> 
> γ Vir is another nice double. For a catalog, this looks promising: http://www.astro.gsu.edu/wds/orb6.html

Damn, I don't have access to WDS catalog anymore  O_o

Yes, adding full support WDS catalog is possible now and probably it should be adding via 2 files - one file for known orbital elements of binaries (multiple systems) and one file for cross identification and informational data (likewise a currect HIP subset of WDS).

Atque commented:
Great work! However, proper motions seem odd now (possibly not calculated for the barycenter of the star system). Compare 61 Cyg in 21,000 AD between this branch and `master`:

![stellarium-315](https://github.com/user-attachments/assets/6c2494c3-3915-41eb-814c-5234aceb10c4)
![stellarium-314](https://github.com/user-attachments/assets/ee646967-4577-4e37-8453-5a6774994a61)

henrysky commented:
> Great work! However, proper motions seem odd now (possibly not calculated for the barycenter of the star system). Compare 61 Cyg in 21,000 AD between this branch and `master`:

Thanks for noticing the inconsistency! Motion of the barycenter is taken into account already but a problem on normalized vector caused the error in position. Now I have fixed the issue and added γ Vir.

Atque commented:
I see some discrepancies with 61 Cyg. Compare the HiPS layer, with the plate in question probably taken around 1990-01-01 (based on the position of 61 Cyg in `master`):

`henrysky-binary` 
![stellarium-317](https://github.com/user-attachments/assets/b3bfb56f-a715-4e3a-8340-06f7ec768651)


`stellarium-master`
![stellarium-318](https://github.com/user-attachments/assets/741e4b7a-c06f-42cd-9723-d14ea678e4f5)

Maybe weird parameters for the stellar orbits?

Atque commented:
I've seen some stars that should be modelled as double stars (I'm not sure if they have defined orbits, though). I simply based this on looking at what stars fly apart in the future or past.

- Alula Australis - ξ UMa - HIP 55203
- Rotanev - β Del - HIP 101769
- Achird - η Cas - HIP 3821
- Mesarthim - γ Ari - HIP 8832
- Mizar - ζ UMa - HIP 65378
- Sheliak - β Lyr - HIP 92420
- ε1 Lyr - HIP 91919
- ε2 Lyr - HIP 91926
- Acamar - θ1 Eri - HIP 13847
- β Tuc - HIP 2484
- Achird - η Cas - HIP 3821
- β Mus - HIP 62322
- Acrux - α1 Cru - HIP 60718
- Dalim - α For - HIP 14879
- γ2 Vol - HIP 34481
- Alya - θ1 Ser - HIP 92946
- γ Lep HIP 27072
- ψ Vel - HIP 46651
- ζ1 Aqr - HIP110960
- ρ Oph - HIP 88601

henrysky commented:
> Maybe weird parameters for the stellar orbits?

I have fixed the parameters so 61 Cyg should match the HiPS layer now.

> I've seen some stars that should be modelled as double stars (I'm not sure if they have defined orbits, though). I simply based this on looking at what stars fly apart in the future or past.

Thanks for the list. I have processed the items till ``Acamar - θ1 Eri - HIP 13847``, I will keep working on the remaining items. As far as I can see, the following binary system won't be added:

- Alula Australis (HIP 55203) - the companion actually does not exist in the catalog
- Mizar (HIP 65378) - there is no Mizar A and B orbital solution only Aa-Ab and Ba-Bb but Mizar Ab and Mizar Bb not resolved in the catalog
- Sheliak (HIP 92420) - there is no β Lyr A and B orbital solution only Aa-Ab but β Lyr Ab not resolved in the catalog
- Acamar (HIP 13847) - there is no orbital solution

Atque commented:
Would it be possible to visualize the orbit, just like planet orbits? And maybe even have markers on it (point of visual/actual apo-/periastron, perhaps markings for dates)?

alex-w commented:
This is maybe helpful:
* δ Sco - [The Revised Orbit of the δ Sco System](https://arxiv.org/abs/1101.5152)
* HD 12376, HD 19971, HD 89795, HD 152027, HD 190412 - [Orbits of Five Triple Stars](https://arxiv.org/abs/2009.10555)
* HIP 14524, HIP 16025, HIP 28671, HIP 46199, HIP 47791, HIP 60444, HIP 61100 and HIP 73085 - [Speckle Interferometry of Nearby Multiple Stars: 2007-2019 Positional Measurements and Orbits of Eight Objects](https://arxiv.org/abs/2005.13958)
* HIP 3645, 14307, 36165, 79980, 103735, 103814, 104440, 105879, 109443 + HIP 41171, 49336, 75663, 78163, and 117666 - [Spectroscopic Orbits of Subsystems in Multiple Stars. IX](https://arxiv.org/abs/2304.01861)
* HIP 14194B, 40523A, 41171A, 51578A, 57572B, 59426A, 62852B, 66438A, 87813B, and 101472A - [Spectroscopic orbits of subsystems in multiple stars. VI](https://arxiv.org/abs/1910.04143)
* [Finding binaries from phase modulation of pulsating stars with Kepler: V. Orbital parameters, with eccentricity and mass-ratio distributions of 341 new binaries](https://arxiv.org/abs/1712.00022)

Maybe adding some catalog of orbital data will be good solution (reading data from this catalog and use it for computations)?

gzotti commented:
> Maybe adding some catalog of orbital data will be good solution (reading data from this catalog and use it for computations)?

Yes, definitely. Even if the list is short now, but inlining every star like that is not elegant. Of course, an extended list means suppressing their "standard" entries in the default star catalog.

henrysky commented:
> > Maybe adding some catalog of orbital data will be good solution (reading data from this catalog and use it for computations)?
> 
> Yes, definitely. Even if the list is short now, but inlining every star like that is not elegant. Of course, an extended list means suppressing their "standard" entries in the default star catalog.

I've made some progress on putting the list in a separate data file! I don't have much time tonight to test but I think the binary stars are still working properly.

gzotti commented:
Great. It really looks fabulous, and I hope this can still go into 25.1 if finished soon. (Needs to be done before RC1!)
The orbit plot idea is also nice but can come later.

Atque commented:
With the latest commits to `stellarium-master`, Zeta Herculis once again has a stationary component. Same with Sabik. So if these parameters are easily found, they ought to be added. Especially Zeta Herculis looks ugly already a few thousand years from the present.

henrysky commented:
Zeta Herculis's and Sabik's companion stars have been dealt with by adding them into the list of binary star system orbital data file.

@alex-w There is a new level 7 catalog here: https://www.astro.utoronto.ca/~hleung/shared/stellarium_new_catalog/1march2025/. Reason: HIP 55987 has been removed from level 7 because it is already included in level 0-4 for being a star with HIP id (even tho it has vmag dimmer than 15).

alex-w commented:
> @alex-w There is a new level 7 catalog here: https://www.astro.utoronto.ca/~hleung/shared/stellarium_new_catalog/1march2025/. Reason: HIP 55987 has been removed from level 7 because it is already included in level 0-4 for being a star with HIP id (even tho it has vmag dimmer than 15).

OK, I've uploading new part of catalog on GH at the moment

P.S. @henrysky, I moved you to core team 2 days ago - you can push fixes and small changes direct to `master` and you can upload parts of catalogs direct on [GH page](https://github.com/Stellarium/stellarium-data/releases/tag/stars-3.0) now

Atque commented:
> Zeta Herculis's and Sabik's companion stars have been dealt with by adding them into the list of binary star system orbital data file.

Neat! I found Castor also has a stationary component. Other than that, I have not found any others.

henrysky commented:
Castor has been added and I think this PR is ready to be reviewed and merged.

There is still one problem where for example, search ``* 61 Cyg`` which points to a custom object instead of the actual star in the catalog (not a new issue but an existing issue). Fixing this by querying children/parents will fix #3181 too but I Don't have much time in the next two weeks so leave that to the future and aim merging this to 25.1

alex-w reviewed (approved):
Great work!

gzotti commented:
I wonder why 61 Cyg B carries the labels like "Bessel's star". Should be either the whole system, or just the brighter component A, right?

gzotti commented on src/core/modules/Star.hpp:
```diff
@@ -243,6 +244,179 @@ struct Star
+      double dec = 0.;
+      getBinaryOrbit(epoch, v, ra, dec, plx, pmra, pmdec, rv);
+   }
+
```
Can you please add some literature documentation here, where you found the algorithms? Maybe Explanatory Supplement (which ed), some other book dedicated on double stars? Or did you do all the math yourself from scratch?

henrysky replied:
I end up just do most of the math myself from definitions of these orbital elements. From the definitions, one can compute the 2D orbit of the secondary star in the reference frame of the primary and then rotate the 2D plane to 3D orbit as seen on the sky and then apply proper motion to the barycenter of the binary star. No fancy algorithms used here.

gzotti replied:
Well, maybe not really fancy and can be done by oneself, but I had expected you followed some textbook, so congratulations.

henrysky replied:
Ah ok, let me google to see if there is any good textbook on those. I forget which resources I use as I indeed have looked up what those orbital elements mean exactly.

gzotti reviewed (approved):
Apart from a bibliography note which I just feel should be there and should not cost much time to add, I won't object a merge. The result just looks awesome, congratulations! When finalizing SUG changes I may edit a bit, but thanks also for these edits and additions.

henrysky commented:
> And is there a way to just "find" 61 Cyg without the custom object?

You can use TAP query on SIMBAD to query parents/children, for example querying ``* 61 Cyg`` children: https://simbad.cds.unistra.fr/simbad/sim-tap/sync?request=doQuery&lang=adql&format=tsv&query=SELECT%20DISTINCT%20basic.main_id,%20basic.ra,%20basic.dec%20FROM%20(SELECT%20oidref%20FROM%20ident%20WHERE%20id=%27*%2061%20Cyg%27)%20AS%20name,%20basic%20JOIN%20h_link%20ON%20basic.oid%20=%20h_link.child%20WHERE%20h_link.parent%20=%20name.oidref which correspond to this query

```
SELECT DISTINCT basic.main_id, basic.ra, basic.dec
FROM (SELECT oidref FROM ident WHERE id='* 61 Cyg') AS name, basic 
JOIN h_link ON basic.oid = h_link.child WHERE h_link.parent = name.oidref
```

And parent of exoplanet ``KOI-4878.01`` by querying https://simbad.cds.unistra.fr/simbad/sim-tap/sync?request=doQuery&lang=adql&format=tsv&query=SELECT%20DISTINCT%20basic.main_id,%20basic.ra,%20basic.dec%20FROM%20(SELECT%20oidref%20FROM%20ident%20WHERE%20id%20=%20%27KOI-4878.01%27)%20AS%20name,%20basic%20JOIN%20h_link%20ON%20basic.oid%20=%20h_link.parent%20WHERE%20h_link.child%20=%20name.oidref which correspond to this query

```
SELECT DISTINCT basic.main_id, basic.ra, basic.dec
FROM (SELECT oidref FROM ident WHERE id = 'KOI-4878.01') AS name, basic 
JOIN h_link ON basic.oid = h_link.parent WHERE h_link.child = name.oidref
```

I don't think it can be

gzotti commented:
I don't mean our casual users should be asked to do SQL queries. They would want to enter "61 Cyg" and expect to find what is now returned only when you know you have to search for "61 Cyg A", i.e., HIP 104214, not a custom object. @alex-w said "Inner dictionary for double stars", but that is still vague. Which file to add "61 Cyg"?

alex-w commented:
> I don't mean our casual users should be asked to do SQL queries. They would want to enter "61 Cyg" and expect to find what is now returned only when you know you have to search for "61 Cyg A", i.e., HIP 104214, not a custom object. @alex-w said "Inner dictionary for double stars", but that is still vague. Which file to add "61 Cyg"?

This file is not exist yet. I've proposed one of solution to resolve objects like multiple stars - select first component when short form of designation is used.

henrysky commented:
> I don't mean our casual users should be asked to do SQL queries.

I meant to say currently Stellarium is using SIMABD scripting interface to query object name (see #3980 that I improve a while ago). So if object name is not in stellarium, it is possible that we use TAP query to check if their parent/child is in Stellarium (I Don't mean for user t do SQL queries themselves).

gzotti commented:
A working but dirty way would be to handle a handful of such commonplace star names like  61 Cyg as proper name in the current skyculture. (I think this was a year ago?)

alex-w commented:
> A working but dirty way would be to handle a handful of such commonplace star names like 61 Cyg as proper name in the current skyculture. (I think this was a year ago?)

Why? Just editing `stars/hip_gaia3/name.fab` file to adding short form of designations for component A

gzotti commented:
OK if that works. Just you said above such "inner dictionary" file does not exist. Is it OK to assign two names ("61 Cyg" and "61 Cyg A") to one star here?

alex-w commented:
> OK if that works. Just you said above such "inner dictionary" file does not exist. Is it OK to assign two names ("61 Cyg" and "61 Cyg A") to one star here?

It is in master already - just check it :)

gzotti commented:
OK, I switched off SIMBAD lookup. But I was lazy, typing "61 cyg" is not enough. "61 Cyg" worked. maybe search should compare any capitalisation here.

alex-w commented:
> OK, I switched off SIMBAD lookup. But I was lazy, typing "61 cyg" is not enough. "61 Cyg" worked. maybe search should compare any capitalisation here.

No. How you will resolve _j Cen_ and _J Cen_ for example?

gzotti commented:
Another find: Compare location of 61 Cyg in 2MASS HiPS. I thought that was taken around 2000? @henrysky is proper motion off? In the DSS, 61Cyg fits to 1975, when the DSS plates are from 1983-2006. 
Groombridge 1830 fits to the Survey dates, also Barnard's and Kapteyn's stars, so maybe something is just off with 61 Cyg's data?

gzotti commented:
> > OK, I switched off SIMBAD lookup. But I was lazy, typing "61 cyg" is not enough. "61 Cyg" worked. maybe search should compare any capitalisation here.
> 
> No. How you will resolve _j Cen_ and _J Cen_ for example?

Good point. You could just offer both to choose from, though.
